### PR TITLE
[WIP] Add logging framework and adapt to logging conventions

### PR DESCRIPTION
**Pull request by wennmo**
Putting up an PR to get some early feedback regarding if this is an approach we want. I have made an attempt at using [logrus](https://github.com/sirupsen/logrus) as logging framework, which seems to be one that is easily customized and fits our purposes. The text and json formatter is in place, and I have updated the test cases.

I have not made any changes yet to the printers for e.g. tables and meta.

Any thoughts about this approach?

Plain text
```bash
➜  corectl git:(log-verbosity) ✗ ./corectl --engine=localhost:9276 build --app test.qvf  --log-level debug
DEBUG[20:51:33] No config file
DEBUG[20:51:33] ---------- Connecting to app ----------
DEBUG[20:51:33] Engine: ws://localhost:9276/app/engineData/ttl/30
DEBUG[20:51:33] SessionId: Q29yZWN0bC1tb3ctU0VMVU4tTU9XLU1BQy10ZXN0LnF2Zg==
DEBUG[20:51:33] OnConnected {"qSessionState":"SESSION_CREATED"}
DEBUG[20:51:33] Document: test.qvf(new)
DEBUG[20:51:33] Meta: http://localhost:9276/v1/apps/test.qvf/data/metadata
INFO[20:51:34] Reload finished successfully
INFO[20:51:34] Saving...
INFO[20:51:34] Done
```

JSON
```bash
➜  corectl git:(log-verbosity) ✗ ./corectl --engine=localhost:9276 build --app test2.qvf  --log-level debug --json
{
  "logseverity": "debug",
  "message": "No config file",
  "timestamp": "2019-03-09T20:51:57+01:00"
}
{
  "logseverity": "debug",
  "message": "---------- Connecting to app ----------",
  "timestamp": "2019-03-09T20:51:57+01:00"
}
{
  "logseverity": "debug",
  "message": "Engine: ws://localhost:9276/app/engineData/ttl/30",
  "timestamp": "2019-03-09T20:51:57+01:00"
}
...
```

JSON + Traffic
```
➜  corectl git:(log-verbosity) ✗ ./corectl --engine=localhost:9276 build --app test2.qvf  --log-level debug --json --traffic
{
  "logseverity": "debug",
  "message": "No config file",
  "timestamp": "2019-03-09T21:00:10+01:00"
}
{
  "logseverity": "debug",
  "message": "---------- Connecting to app ----------",
  "timestamp": "2019-03-09T21:00:10+01:00"
}
{
  "logseverity": "debug",
  "message": "Engine: ws://localhost:9276/app/engineData/ttl/30",
  "timestamp": "2019-03-09T21:00:10+01:00"
}
{
  "logseverity": "debug",
  "message": "SessionId: Q29yZWN0bC1tb3ctU0VMVU4tTU9XLU1BQy10ZXN0Mi5xdmY=",
  "timestamp": "2019-03-09T21:00:10+01:00"
}
{
  "direction": "inbound",
  "logseverity": "info",
  "message": "{\"jsonrpc\":\"2.0\",\"method\":\"OnConnected\",\"params\":{\"qSessionState\":\"SESSION_CREATED\"}}",
  "timestamp": "2019-03-09T21:00:10+01:00",
  "type": "traffic"
}
{
  "logseverity": "debug",
  "message": "OnConnected {\"qSessionState\":\"SESSION_CREATED\"}",
  "timestamp": "2019-03-09T21:00:10+01:00"
}
{
  "direction": "outbound",
  "logseverity": "info",
  "message": "{\"jsonrpc\":\"2.0\",\"delta\":false,\"method\":\"GetActiveDoc\",\"handle\":-1,\"id\":1,\"params\":[]}",
  "timestamp": "2019-03-09T21:00:10+01:00",
  "type": "traffic"
}
{
  "direction": "inbound",
  "logseverity": "info",
  "message": "{\"jsonrpc\":\"2.0\",\"id\":1,\"error\":{\"code\":1007,\"parameter\":\"No active document\",\"message\":\"App invalid\"}}",
  "timestamp": "2019-03-09T21:00:10+01:00",
  "type": "traffic"
}
{
  "direction": "outbound",
  "logseverity": "info",
  "message": "{\"jsonrpc\":\"2.0\",\"delta\":false,\"method\":\"OpenDoc\",\"handle\":-1,\"id\":2,\"params\":[\"test2.qvf\",\"\",\"\",\"\",false]}",
  "timestamp": "2019-03-09T21:00:10+01:00",
  "type": "traffic"
}
...
```

### CODE REVIEW

File: docs/corectl.md [hunk 1]
```diff
@@ -13,9 +13,10 @@ corectl [flags]
 ### Options
 
 ```
-  -h, --help      help for corectl
-  -t, --traffic   Log JSON websocket traffic to stdout
-  -v, --verbose   Logs extra information
+  -h, --help               help for corectl
+  -j, --json               Set logging format to JSON
+  -l, --log-level string   Set logging level, one of; TRACE, DEBUG, INFO, WARN, ERROR, FATAL and PANIC (default "INFO")
+  -t, --traffic            Log JSON websocket traffic to stdout
```
Comment (peol) [inline docs/corectl.md]: Maybe we can include traffic if log level is debug? One less flag to keep track of that way.

File: internal/modeldata.go [hunk 2]
```diff
@@ -125,7 +122,7 @@ func GetModelMetadata(ctx context.Context, doc *enigma.Doc, metaURL string, head
 	restMetadata, err := ReadRestMetadata(metaURL, headers)
 
 	if len(tables) > 0 && restMetadata == nil {
-		fmt.Println("No REST metadata available.")
+		Logger.Info("No REST metadata available.")
```
Comment (peol) [inline internal/modeldata.go]: warn?

File: docs/corectl.md (hunk 1 above)
Comment (wennmo) [reply docs/corectl.md]: Good point 👍

**Comment by wennmo** [comment]
Updated version where traffic is included in log level DEBUG, and where traffic is rendered as real JSON

Plain text + debug (now including traffic)
```
➜  corectl git:(log-verbosity) ✗ ./corectl --engine=localhost:9276 build --app test2.qvf --log-level=DEBUG
DEBUG[13:25:21] ---------- Connecting to app ----------      
DEBUG[13:25:21] Engine: ws://localhost:9276/app/engineData/ttl/30 
DEBUG[13:25:21] SessionId: Q29yZWN0bC1tb3ctU0VMVU4tTU9XLU1BQy5sb2NhbC10ZXN0Mi5xdmY= 
DEBUG[13:25:21] {"jsonrpc":"2.0","method":"OnConnected","params":{"qSessionState":"SESSION_CREATED"}}  type=traffic direction=inbound
DEBUG[13:25:21] OnConnected {"qSessionState":"SESSION_CREATED"} 
DEBUG[13:25:21] {"jsonrpc":"2.0","delta":false,"method":"GetActiveDoc","handle":-1,"id":1,"params":[]}  type=traffic direction=outbound
DEBUG[13:25:21] {"jsonrpc":"2.0","id":1,"error":{"code":1007,"parameter":"No active document","message":"App invalid"}}  type=traffic direction=inbound
DEBUG[13:25:21] {"jsonrpc":"2.0","delta":false,"method":"OpenDoc","handle":-1,"id":2,"params":["test2.qvf","","","",false]}  type=traffic direction=outbound
DEBUG[13:25:21] {"jsonrpc":"2.0","id":2,"result":{"qReturn":{"qType":"Doc","qHandle":1,"qGenericId":"/apps/test2.qvf"}},"change":[1]}  type=traffic direction=inbound
DEBUG[13:25:21] App:  test2.qvf(opened)                      
DEBUG[13:25:21] Meta: http://localhost:9276/v1/apps/test2.qvf/data/metadata
...
````

JSON + debug
```
➜  corectl git:(log-verbosity) ✗ ./corectl --engine=localhost:9276 build --app test2.qvf --log-level=DEBUG --json
{"logseverity":"debug","message":"No config file","timestamp":"2019-03-11T13:46:47+01:00"}
{"logseverity":"debug","message":"---------- Connecting to app ----------","timestamp":"2019-03-11T13:46:47+01:00"}
{"logseverity":"debug","message":"Engine: ws://localhost:9276/app/engineData/ttl/30","timestamp":"2019-03-11T13:46:47+01:00"}
{"logseverity":"debug","message":"SessionId: Q29yZWN0bC1tb3ctU0VMVU4tTU9XLU1BQy5sb2NhbC10ZXN0Mi5xdmY=","timestamp":"2019-03-11T13:46:47+01:00"}
{"logseverity":"debug","message":{"jsonrpc":"2.0","method":"OnConnected","params":{"qSessionState":"SESSION_CREATED"}},"type":"traffic","direction":"inbound","timestamp":"2019-03-11 12:46:47.783028 +0000 UTC"}
{"logseverity":"debug","message":"OnConnected {\"qSessionState\":\"SESSION_CREATED\"}","timestamp":"2019-03-11T13:46:47+01:00"}
{"logseverity":"debug","message":{"jsonrpc":"2.0","delta":false,"method":"GetActiveDoc","handle":-1,"id":1,"params":[]},"type":"traffic","direction":"outbound","timestamp":"2019-03-11 12:46:47.783316 +0000 UTC"}
{"logseverity":"debug","message":{"jsonrpc":"2.0","id":1,"error":{"code":1007,"parameter":"No active document","message":"App invalid"}},"type":"traffic","direction":"inbound","timestamp":"2019-03-11 12:46:47.785325 +0000 UTC"}
...
```

**Comment by peol** [comment]
Is it a problem if `message` can be both a string and an object? Maybe it's better to keep the message as string and a new property for the traffic data?

**Comment by wennmo** [comment]
@peol Yeah, that might become a problem when parsing the logs. Perhaps we should have an empty `message`, put the traffic entry in `traffic`, and skip the `type`?

I also saw that I need to fix the timestamps on the traffic log entry :)

**Comment by peol** [comment]
@wennmo Sounds good, but I think `message` could be like `sent` or `recv` if it's traffic?

**Comment by wennmo** [comment]
@peol Ok great, I can add that. I'll remove the `direction` key as well then.

**Comment by wennmo** [comment]
@peol
When logging traffic entries those are populated in `traffic` when using `--json` flag. For plain text the traffic will be printed in `message` to be more human readable.
Took a look at kitlog and glog as well but they seem less customizable when it comes to plain text, so using logrus is still the preferred choice in my opinion.

Plain text (TTY) with log-level debug:
```
➜  corectl git:(log-verbosity) ✗ ./corectl --engine=localhost:9276 build --app test2.qvf --log-level=debug             
DEBUG[14:26:33] No config file                               
DEBUG[14:26:33] ---------- Connecting to app ----------      
DEBUG[14:26:33] Engine: ws://localhost:9276/app/engineData/ttl/30 
DEBUG[14:26:33] SessionId: Q29yZWN0bC1tb3ctU0VMVU4tTU9XLU1BQy5sb2NhbC10ZXN0Mi5xdmY= 
DEBUG[14:26:33] {"jsonrpc":"2.0","method":"OnConnected","params":{"qSessionState":"SESSION_ATTACHED"}}  traffic=recv
DEBUG[14:26:33] OnConnected {"qSessionState":"SESSION_ATTACHED"} 
DEBUG[14:26:33] {"jsonrpc":"2.0","delta":false,"method":"GetActiveDoc","handle":-1,"id":1,"params":[]}  traffic=sent
DEBUG[14:26:33] {"jsonrpc":"2.0","id":1,"result":{"qReturn":{"qType":"Doc","qHandle":1,"qGenericId":"/apps/test2.qvf"}}}  traffic=recv
```

Plain text without TTY:
```
➜  corectl git:(log-verbosity) ✗ ./corectl --engine=localhost:9276 build --app test2.qvf --log-level=debug
timestamp="14:26:17" logseverity=debug message="No config file"
timestamp="14:26:17" logseverity=debug message="---------- Connecting to app ----------"
timestamp="14:26:17" logseverity=debug message="Engine: ws://localhost:9276/app/engineData/ttl/30"
timestamp="14:26:17" logseverity=debug message="SessionId: Q29yZWN0bC1tb3ctU0VMVU4tTU9XLU1BQy5sb2NhbC10ZXN0Mi5xdmY="
timestamp="14:26:17" logseverity=debug message="{\"jsonrpc\":\"2.0\",\"method\":\"OnConnected\",\"params\":{\"qSessionState\":\"SESSION_CREATED\"}}" traffic=recv
timestamp="14:26:17" logseverity=debug message="OnConnected {\"qSessionState\":\"SESSION_CREATED\"}"
timestamp="14:26:17" logseverity=debug message="{\"jsonrpc\":\"2.0\",\"delta\":false,\"method\":\"GetActiveDoc\",\"handle\":-1,\"id\":1,\"params\":[]}" traffic=sent
timestamp="14:26:18" logseverity=debug message="{\"jsonrpc\":\"2.0\",\"id\":1,\"error\":{\"code\":1007,\"parameter\":\"No active document\",\"message\":\"App invalid\"}}" traffic=recv
timestamp="14:26:18" logseverity=debug message="{\"jsonrpc\":\"2.0\",\"delta\":false,\"method\":\"OpenDoc\",\"handle\":-1,\"id\":2,\"params\":[\"test2.qvf\",\"\",\"\",\"\",false]}" traffic=sent
timestamp="14:26:18" logseverity=debug message="{\"jsonrpc\":\"2.0\",\"id\":2,\"result\":{\"qReturn\":{\"qType\":\"Doc\",\"qHandle\":1,\"qGenericId\":\"/apps/test2.qvf\"}},\"change\":[1]}" traffic=recv
timestamp="14:26:18" logseverity=debug message="App:  test2.qvf(opened)"
```

JSON with log-level debug:
```
➜  corectl git:(log-verbosity) ✗ ./corectl --engine=localhost:9276 build --app test2.qvf --log-level=debug --json
{"logseverity":"debug","message":"No config file","timestamp":"2019-03-12T14:29:51+01:00"}
{"logseverity":"debug","message":"---------- Connecting to app ----------","timestamp":"2019-03-12T14:29:51+01:00"}
{"logseverity":"debug","message":"Engine: ws://localhost:9276/app/engineData/ttl/30","timestamp":"2019-03-12T14:29:51+01:00"}
{"logseverity":"debug","message":"SessionId: Q29yZWN0bC1tb3ctU0VMVU4tTU9XLU1BQy5sb2NhbC10ZXN0Mi5xdmY=","timestamp":"2019-03-12T14:29:51+01:00"}
{"logseverity":"debug","message":"recv","traffic":{"jsonrpc":"2.0","method":"OnConnected","params":{"qSessionState":"SESSION_CREATED"}},"timestamp":"2019-03-12T14:29:51+01:00"}
{"logseverity":"debug","message":"OnConnected {\"qSessionState\":\"SESSION_CREATED\"}","timestamp":"2019-03-12T14:29:51+01:00"}
{"logseverity":"debug","message":"sent","traffic":{"jsonrpc":"2.0","delta":false,"method":"GetActiveDoc","handle":-1,"id":1,"params":[]},"timestamp":"2019-03-12T14:29:51+01:00"}
{"logseverity":"debug","message":"recv","traffic":{"jsonrpc":"2.0","id":1,"error":{"code":1007,"parameter":"No active document","message":"App invalid"}},"timestamp":"2019-03-12T14:29:51+01:00"}
{"logseverity":"debug","message":"sent","traffic":{"jsonrpc":"2.0","delta":false,"method":"OpenDoc","handle":-1,"id":2,"params":["test2.qvf","","","",false]},"timestamp":"2019-03-12T14:29:51+01:00"}
{"logseverity":"debug","message":"recv","traffic":{"jsonrpc":"2.0","id":2,"result":{"qReturn":{"qType":"Doc","qHandle":1,"qGenericId":"/apps/test2.qvf"}},"change":[1]},"timestamp":"2019-03-12T14:29:51+01:00"}
{"logseverity":"debug","message":"App:  test2.qvf(opened)","timestamp":"2019-03-12T14:29:51+01:00"}
{"logseverity":"debug","message":"Meta: http://localhost:9276/v1/apps/test2.qvf/data/metadata","timestamp":"2019-03-12T14:29:51+01:00"}
{"logseverity":"debug","message":"sent","traffic":{"jsonrpc":"2.0","delta":false,"method":"GetConnections","handle":1,"id":3,"params":[]},"timestamp":"2019-03-12T14:29:51+01:00"}
{"logseverity":"debug","message":"recv","traffic":{"jsonrpc":"2.0","id":3,"result":{"qConnections":[]}},"timestamp":"2019-03-12T14:29:51+01:00"}
{"logseverity":"debug","message":"sent","traffic":{"jsonrpc":"2.0","delta":false,"method":"DoReload","handle":1,"id":4,"params":[0,false,false]},"timestamp":"2019-03-12T14:29:51+01:00"}
```

**Comment by wennmo** [comment]
Closing PR for now. After discussion this is not the approach we want since it has a too big impact on plain text output

**Pull request closed without merging**